lib/libmp/src/core/mp_bus.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "mp_bus.h"
+
+typedef enum {
+	MP_BUS_DROP = 0,
+	MP_BUS_PASS = 1,
+} MpBusSyncReply;
+
+/**
+ * Sync handler for the bus.
+ * This function is called when a message is posted to the bus.
+ * It delivers the message to the correct listener based on the filter type.
+ * If a listener consumes the message, it returns MP_BUS_DROP,
+ * otherwise it returns MP_BUS_PASS.
+ *
+ * @param bus: a @ref MpBus to handle the message
+ * @param message: the message to handle
+ */
+static MpBusSyncReply mp_bus_sync_handler(MpBus *bus, MpMessage *message)
+{
+	MpBusSyncListener *listener;
+	bool ret = false;
+
+	/* Deliver the message to the correct listener */
+	SYS_SLIST_FOR_EACH_CONTAINER(&bus->sync_listeners, listener, node) {
+		if (MP_MESSAGE_TYPE(message) & listener->filter_type) {
+			ret |= listener->callback(message, listener->user_data);
+		}
+	}
+
+	return ret ? MP_BUS_DROP : MP_BUS_PASS;
```

Comment:
If this library is meant to be used with Zephyr only, then it could become standard `-ERRNO` error codes.

However, that is also getting in your way as `ret |= ...` would no be possible anymore (unless `return ret == 0 ? 0 : -ESOMETHING` is used).